File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        last = head
        res = head
        if head is None:
            return None
        while last is not None and n > 0:
            last = last.next
            n -= 1
        prev = None
        removed_node = head
        while last is not None:
            prev = removed_node
            removed_node = removed_node.next
            last = last.next
        if prev is None:
            return res.next
        if removed_node is not None:
            prev.next = removed_node.next
        return res

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* curr = head;
        
        while (n != 0) {
            curr = curr->next;
            n--;
        }
        ListNode* front = head;
        ListNode* prev = NULL;
        
        if (!curr) return head->next;
        
        
        while (curr) {
            prev = front;
            front = front->next;
            curr = curr->next;
        }
        if (prev) {
            prev->next = front->next;
        }
        else {
            return prev;
        }
        return head;
    }
};